src/main/java/org/altbeacon/beacon/service/GattBeaconTracker.java

```diff
@@ -69,13 +62,12 @@ private Beacon trackGattBeacon(Beacon beacon) {
     private void updateTrackingHashes(Beacon trackedBeacon, HashMap<Integer,Beacon> matchingTrackedBeacons) {
         if (matchingTrackedBeacons == null) {
             matchingTrackedBeacons = new HashMap<Integer,Beacon>();
-            mBeaconsByServiceUuidAndMac.put(serviceUuidAndMac(trackedBeacon), matchingTrackedBeacons);
         }
-        mBeacons.put(trackedBeacon.hashCode(), trackedBeacon);
         matchingTrackedBeacons.put(trackedBeacon.hashCode(), trackedBeacon);
+        mBeaconsByKey.put(getBeaconKey(trackedBeacon), matchingTrackedBeacons);
     }
 
-    private String serviceUuidAndMac(Beacon beacon) {
-        return beacon.getBluetoothAddress()+beacon.getServiceUuid();
+    private String getBeaconKey(Beacon beacon) {
+        return beacon.getBluetoothAddress() + beacon.getServiceUuid();
```

Comment:
What if the serviceUUID is not -1, but `beacon.isMultiFrameBeacon()` returns true?  Wouldn't this prevent associating extra data between beacons with different serviceUUIDs but the same Mac?  Is this needed in order to ensure backward compatibility?